containers/source-build/source_build.py

```diff
@@ -0,0 +1,455 @@
+#!/usr/bin/python3.11
+
+import argparse
+import functools
+import json
+import itertools
+import os
+import shutil
+import logging
+import stat
+import sys
+import tarfile
+import tempfile
+import urllib.parse
+from dataclasses import dataclass, field
+from subprocess import PIPE, run, CalledProcessError
+from typing import TypedDict, NotRequired, Literal, Final
+
+"""
+Requires: git, skopeo, tar, BuildSourceImage
+"""
+
+logging.basicConfig(level=logging.DEBUG, format="%(asctime)s:%(name)s:%(levelname)s:%(message)s")
+logger = logging.getLogger("source-build")
+
+BSI: Final = "/opt/BuildSourceImage/bsi"
+
+
+class BuildResult(TypedDict):
+    status: Literal["failure", "success"]
+    message: NotRequired[str]
+    dependencies_included: bool
+    base_image_source_included: bool
+    image_url: str
+
+
+@dataclass
+class SourceImageBuildDirectories:
+    rpm_dir: str = ""
+    extra_src_dirs: list[str] = field(default_factory=list)
+
+
+class RepoInfo(TypedDict):
+    name: str
+    last_commit_sha: str
+
+
+def arg_type_path(value):
+    if not os.path.exists(value):
+        raise argparse.ArgumentTypeError(f"No file or directory exists at specified path {value}")
+    return value
+
+
+def arg_type_bsi_script(value):
+    if not os.path.exists(value):
+        raise argparse.ArgumentTypeError(f"BuildSourceImage executable {value} does not exist")
+    fstat = os.stat(value)
+    if not stat.S_IXUSR & fstat.st_mode:
+        raise argparse.ArgumentTypeError(f"BuildSourceImage script {value} is not executable")
+    return value
+
+
+def arg_type_base_images(value):
+    return value.strip()
+
+
+def get_repo_info(repo_path: str) -> RepoInfo:
+    _run = functools.partial(run, check=True, text=True, capture_output=True, cwd=repo_path)
+    commit_sha = _run(["git", "rev-parse", "HEAD"]).stdout.strip()
+    repo_url = _run(["git", "config", "--get", "remote.origin.url"]).stdout.strip()
+    # The url could look like https://github.com/namespace/app.git
+    # where, app is required.
+    repo_name, _ = os.path.splitext(repo_url.rsplit("/", maxsplit=1)[-1])
```

Comment:
> Should we remove the .git suffix from the repo name?

I'm not sure I get your point. `os.path.splitext` is called.

> Should we handle urls with trailing /? E.g. https://github.com/namespace/app.git/

This is got from the repo cloned by git-clone. Could `remote.origin.url` return such a url?

Reply:
Ah, my bad, I missed the splitext

> This is got from the repo cloned by git-clone. Could remote.origin.url return such a url?

Yeah, it can
```bash
$ git clone git@github.com:containerbuildsystem/cachi2.git/
$ cd cachi2
$ git config --get remote.origin.url
containerbuildsystem/cachi2.git/
```